crates/maybenot/src/counter.rs

```diff
@@ -9,62 +9,92 @@ use std::fmt;
-/// distribution to sample values from when updating the counter.
+/// A specification of how one of a [`Machine`]'s counters should be updated
+/// when transitioning to a [`State`](crate::state::State). Consists of an
+/// [`Operation`] to be applied to the counter with one of three values: by
+/// default, the value 1, unless a distribution is provided or the copy flag is
+/// set to true. If the copy flag is set to true, the counter will be updated
+/// with the value of the other counter *prior to transitioning to the state*.
+/// If a distribution is provided, the counter will be updated with a value
+/// sampled from the distribution.
 #[derive(PartialEq, Debug, Clone, Copy, Serialize, Deserialize)]
-pub struct CounterUpdate {
-    /// Which counter to update.
-    pub counter: Counter,
-    /// The operation to apply to the counter upon a state transition.
+pub struct Counter {
+    /// The operation to apply to the counter upon a state transition. If the
+    /// distribution is not set and copy is false, the counter will be updated
+    /// by 1.
     pub operation: Operation,
     /// If set, sample the value to update the counter with from a
-    /// distribution. If not set, value is 1.
-    pub value: Option<Dist>,
+    /// distribution.
+    pub dist: Option<Dist>,
+    /// If set, the counter will be updated by the other counter's value *prior
+    /// to transitioning to the state*. Supersedes the `dist` field.
+    pub copy: bool,
 }
 
-impl fmt::Display for CounterUpdate {
+impl fmt::Display for Counter {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         write!(f, "{:#?}", self)
     }
 }
 
-impl CounterUpdate {
+impl Counter {
+    /// Create a new counter with an operation that modifies the counter with
+    /// value 1.
+    pub fn new(operation: Operation) -> Self {
+        Counter {
+            operation,
+            dist: None,
+            copy: false,
+        }
+    }
+
+    /// Create a new counter with an operation and a distribution to sample the
+    /// value from.
+    pub fn new_dist(operation: Operation, dist: Dist) -> Self {
+        Counter {
+            operation,
+            dist: Some(dist),
+            copy: false,
+        }
+    }
+
+    /// Create a new counter with an operation that copies the value of the
+    /// other counter *prior to transitioning to the state*.
+    pub fn new_copy(operation: Operation) -> Self {
+        Counter {
+            operation,
+            dist: None,
+            copy: true,
+        }
+    }
+
     /// Sample a value to update the counter with.
     pub fn sample_value<R: RngCore>(&self, rng: &mut R) -> u64 {
-        match self.value {
-            Some(value) => {
-                let s = value.sample(rng) as u64;
+        match self.dist {
+            None => 1,
+            Some(dist) => {
+                let s = dist.sample(rng) as u64;
                 s.min(MAX_SAMPLED_COUNTER_VALUE)
```

Comment:
This will not yield all values in the range `0 - MAX_SAMPLED_COUNTER_VALUE`. But I guess you know that. This part of the machine is not something I know very much about.

Reply:
I guess you're referring to the rounding down of the int part of `f64` + size difference? Ended up cutting the constant now, the converting from samples `f64` to `u64` is fine for all possible use-cases of the framework right now.